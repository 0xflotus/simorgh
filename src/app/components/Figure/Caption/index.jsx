import React, { Fragment } from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import {
  FF_NEWS_SANS_REG,
  C_STORM,
  GEL_SPACING,
} from '../../../lib/constants/styles';
import VisuallyHiddenText from '../../VisuallyHiddenText';
import mediaQuery from '../../../helpers/mediaQueries';

const StyledCaption = styled.figcaption`
  background-color: #d5d0cd;
  color: ${C_STORM};
  font-family: ${FF_NEWS_SANS_REG};
  padding: ${GEL_SPACING};

  // Font styling below is a subset of BBC GEL Typography "Long Primer"
  font-size: 0.9375em;
  line-height: 1.125rem;
  ${mediaQuery.laptopAndLarger} {
    font-size: 0.875em;
  }
`;

const Caption = ({ text }) => (
  <Fragment>
    <VisuallyHiddenText>Image caption, </VisuallyHiddenText>
    <StyledCaption>{text}</StyledCaption>
  </Fragment>
);

Caption.propTypes = {
  text: string.isRequired,
};

export default Caption;
