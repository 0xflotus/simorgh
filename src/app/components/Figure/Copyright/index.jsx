import React, { Fragment } from 'react';
import { node } from 'prop-types';
import styled from 'styled-components';
import {
  FF_NEWS_SANS_REG,
  C_WHITE,
  GEL_SPACING,
  GEL_SPACING_HLF,
} from '../../../lib/constants/styles';
import { T_MINION } from '../../../lib/constants/typography';
import VisuallyHiddenText from '../../VisuallyHiddenText';

const StyledCopyright = styled.span`
  ${T_MINION};
  background-color: rgba(34, 34, 34, 0.75);
  text-transform: uppercase;
  color: ${C_WHITE};
  padding: ${GEL_SPACING_HLF} ${GEL_SPACING};
  font-family: ${FF_NEWS_SANS_REG};
  position: absolute;
  bottom: 0;
  right: 0;
`;

const Copyright = ({ children }) => (
  <Fragment>
    <VisuallyHiddenText>Image source, {children}</VisuallyHiddenText>
    <StyledCopyright aria-hidden="true">{children}</StyledCopyright>
  </Fragment>
);

Copyright.propTypes = {
  children: node.isRequired,
};

export default Copyright;
