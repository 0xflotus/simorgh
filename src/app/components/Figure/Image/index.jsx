import React from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';

const StyledImage = styled.img`
  display: block;
  width: 100%;
`;

const Image = ({ src, alt }) => <StyledImage alt={alt} src={src} />;

Image.propTypes = {
  alt: string.isRequired,
  src: string.isRequired,
};

export default Image;
