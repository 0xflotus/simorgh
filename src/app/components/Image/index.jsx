import React from 'react';
import propTypes from 'prop-types';

const getAltText = block => block.model.blocks[0].model.blocks[0].model.text;

const Image = ({ blocks }) => {
  const subBlocks = blocks[0].model.blocks;
  const rawImageBlock = subBlocks.filter(block => block.type === 'rawImage')[0];
  const altTextBlock = subBlocks.filter(block => block.type === 'altText')[0];

  if (!rawImageBlock) {
    return null;
  }

  const rawImageLocator = rawImageBlock.model.locator;
  const altText = altTextBlock ? getAltText(altTextBlock) : '';
  const rawImageSrc = `https://ichef.bbci.co.uk/news/640${rawImageLocator}`;

  return <img alt={altText} src={rawImageSrc} />;
};

Image.propTypes = {
  blocks: propTypes.arrayOf(
    propTypes.shape({
      model: propTypes.shape({
        blocks: propTypes.arrayOf(
          propTypes.shape({
            locator: propTypes.string,
          }),
        ),
      }),
    }),
  ),
};

Image.defaultProps = {
  blocks: [
    {
      model: {
        blocks: [
          {
            model: {},
          },
        ],
      },
    },
  ],
};

export default Image;
