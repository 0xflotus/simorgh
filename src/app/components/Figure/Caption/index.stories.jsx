import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import VisuallyHiddenText from '../../VisuallyHiddenText';
import Caption from './index';

storiesOf('Caption', module).add('default', () => (
  <Caption>
    <VisuallyHiddenText>Image caption,</VisuallyHiddenText>
    This is some Caption text
  </Caption>
));
