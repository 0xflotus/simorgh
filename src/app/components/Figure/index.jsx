import React from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import Caption from './Caption';

const renderCaption = caption =>
  caption ? <Caption>{caption}</Caption> : null;

const StyledFigure = styled.figure`
  padding-bottom: 16px;
  width: 100%;
`;

const Image = styled.img.attrs({
  src: ({ src }) => src,
  alt: ({ alt }) => alt,
})`
  width: 100%;
`;

const Figure = ({ src, alt, caption }) => (
  <StyledFigure>
    <Image src={src} alt={alt} />
    {renderCaption(caption)}
  </StyledFigure>
);

Figure.propTypes = {
  alt: string.isRequired,
  src: string.isRequired,
  caption: string,
};

Figure.defaultProps = {
  caption: null,
};

export default Figure;
