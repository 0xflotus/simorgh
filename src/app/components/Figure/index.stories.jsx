import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import Figure from './index';
import Caption from './Caption';
import VisuallyHiddenText from '../VisuallyHiddenText';

const image = {
  alt:
    'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.',
  src:
    'https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png',
};

const caption = 'This is a caption!';
const copyright = 'Copyright Getty images';

storiesOf('Figure', module)
  .add('with a caption', () => (
    <Figure {...image}>
      <Caption>{caption}</Caption>
    </Figure>
  ))
  .add('without a caption', () => <Figure {...image} />)
  .add('with non-BBC copyright', () => (
    <Figure {...image}>
      <VisuallyHiddenText>{copyright}</VisuallyHiddenText>
    </Figure>
  ))
  .add('with a caption and non-BBC copyright', () => (
    <Figure {...image}>
      <VisuallyHiddenText>{copyright}</VisuallyHiddenText>
      <Caption>{caption}</Caption>
    </Figure>
  ));
