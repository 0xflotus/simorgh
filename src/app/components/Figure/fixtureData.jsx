import React from 'react';
import Figure from './index';
import Image from './Image';
import Caption from './Caption';
import Text from '../Text';
import VisuallyHiddenText from '../VisuallyHiddenText';

const imageAlt =
  'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.';
const imageSrc =
  'https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png';
const captionValue = 'This is a caption';
const captionValueContainingLink =
  'This is a caption [with a link](https://bbc.com/news)';
const copyrightText = 'Copyright Getty images';

export const FigureImage = (
  <Figure>
    <Image alt={imageAlt} src={imageSrc} />
  </Figure>
);

export const FigureImageWithCaption = (
  <Figure>
    <Image alt={imageAlt} src={imageSrc} />
    <Text text={captionValue} paragraphOverride={Caption} />
  </Figure>
);

export const FigureImageWithCaptionContainingLink = (
  <Figure>
    <Image alt={imageAlt} src={imageSrc} />
    <Text text={captionValueContainingLink} paragraphOverride={Caption} />
  </Figure>
);

export const FigureImageWithCopyright = (
  <Figure>
    <Image alt={imageAlt} src={imageSrc} />
    <VisuallyHiddenText>{copyrightText}</VisuallyHiddenText>
  </Figure>
);

export const FigureImageWithCopyrightAndCaption = (
  <Figure>
    <Image alt={imageAlt} src={imageSrc} />
    <VisuallyHiddenText>{copyrightText}</VisuallyHiddenText>
    <Text text={captionValue} paragraphOverride={Caption} />
  </Figure>
);
