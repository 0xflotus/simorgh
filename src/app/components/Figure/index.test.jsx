import React from 'react';
import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import Figure from './index';
import Image from './Image';
import Caption from './Caption';
import VisuallyHiddenText from '../VisuallyHiddenText';

const imageAlt =
  'Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17.';
const imageSrc =
  'https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png';
const captionValue = 'This is a caption';
const copyrightText = 'Copyright Getty images';

describe('Figure', () => {
  describe('with a caption', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure>
        <Image alt={imageAlt} src={imageSrc} />
        <Caption>{captionValue}</Caption>
      </Figure>,
    );
  });

  describe('without a caption', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure>
        <Image alt={imageAlt} src={imageSrc} />
      </Figure>,
    );
  });

  describe('with non-BBC copyright', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure>
        <Image alt={imageAlt} src={imageSrc} />
        <VisuallyHiddenText>{copyrightText}</VisuallyHiddenText>
      </Figure>,
    );
  });

  describe('with caption and non-BBC copyright', () => {
    shouldMatchSnapshot(
      'should render correctly',
      <Figure>
        <Image alt={imageAlt} src={imageSrc} />
        <VisuallyHiddenText>{copyrightText}</VisuallyHiddenText>
        <Caption>{captionValue}</Caption>
      </Figure>,
    );
  });
});
