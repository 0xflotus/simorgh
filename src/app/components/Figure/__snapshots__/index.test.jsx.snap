// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure with a caption should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c1 {
  display: block;
  width: 100%;
}

<figure
  className="c0"
>
  <img
    alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
    className="c1"
    src="https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png"
  />
</figure>
`;

exports[`Figure with caption and non-BBC copyright should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c1 {
  display: block;
  width: 100%;
}

<figure
  className="c0"
>
  <img
    alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
    className="c1"
    src="https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png"
  />
</figure>
`;

exports[`Figure with non-BBC copyright should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c1 {
  display: block;
  width: 100%;
}

<figure
  className="c0"
>
  <img
    alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
    className="c1"
    src="https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png"
  />
</figure>
`;

exports[`Figure without a caption should render correctly 1`] = `
.c0 {
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c1 {
  display: block;
  width: 100%;
}

<figure
  className="c0"
>
  <img
    alt="Map of the UK displaying Syrian refugees and asylum seekers per 10000 population. Ranges from 0 to 17."
    className="c1"
    src="https://ichef.bbci.co.uk/news/640/cpsprodpb/439A/production/_100960371_syrians_and_asylum_v2-nc.png"
  />
</figure>
`;
